src/uthash.h

```diff
@@ -177,24 +197,45 @@ do {
     uthash_fatal("out of memory");                                               \
   }                                                                              \
-  uthash_bzero((head)->hh.tbl, sizeof(UT_hash_table));                           \
-  (head)->hh.tbl->tail = &((head)->hh);                                          \
-  (head)->hh.tbl->num_buckets = HASH_INITIAL_NUM_BUCKETS;                        \
-  (head)->hh.tbl->log2_num_buckets = HASH_INITIAL_NUM_BUCKETS_LOG2;              \
-  (head)->hh.tbl->hho = (char*)(&(head)->hh) - (char*)(head);                    \
-  (head)->hh.tbl->buckets = (UT_hash_bucket*)uthash_malloc(                      \
-      HASH_INITIAL_NUM_BUCKETS * sizeof(struct UT_hash_bucket));                 \
-  if (!(head)->hh.tbl->buckets) {                                                \
-    uthash_fatal("out of memory");                                               \
+  do {                                                                           \
+    uthash_bzero((head)->hh.tbl, sizeof(UT_hash_table));                         \
+    (head)->hh.tbl->tail = &((head)->hh);                                        \
+    (head)->hh.tbl->num_buckets = HASH_INITIAL_NUM_BUCKETS;                      \
+    (head)->hh.tbl->log2_num_buckets = HASH_INITIAL_NUM_BUCKETS_LOG2;            \
+    (head)->hh.tbl->hho = (char*)(&(head)->hh) - (char*)(head);                  \
+    (head)->hh.tbl->buckets = (UT_hash_bucket*)uthash_malloc(                    \
+        HASH_INITIAL_NUM_BUCKETS * sizeof(struct UT_hash_bucket));               \
+    if (!(head)->hh.tbl->buckets) {                                              \
+      if (HASH_NOMEM_OK) {                                                       \
+        _mem_ok = 0;                                                             \
+        break;                                                                   \
+      }                                                                          \
+      uthash_fatal("out of memory");                                             \
+    }                                                                            \
+    uthash_bzero((head)->hh.tbl->buckets,                                        \
+        HASH_INITIAL_NUM_BUCKETS * sizeof(struct UT_hash_bucket));               \
+    HASH_BLOOM_MAKE((head)->hh.tbl,_mem_ok);                                     \
+    if (HASH_NOMEM_OK && !_mem_ok) {                                             \
+      break;                                                                     \
+    }                                                                            \
+    (head)->hh.tbl->signature = HASH_SIGNATURE;                                  \
+  } while (0);                                                                   \
+  if (HASH_NOMEM_OK && !_mem_ok) {                                               \
```

Comment:
I'm not a fan of this repeated `if` test, for a condition that most of the time will be false, and if it is true, we knew it was true when we jumped here. Can you refactor this to get rid of the test? (Obviously it would be easy with a `goto`, but I think we have to avoid `goto` also. ;))

This is also a case of `if (HASH_NOMEM_OK ...` which should be handled with a `#define HASH_MAKE_TABLE_OOM(...)` somewhere up top.